File: Day 4/solution.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>
#include <cassert>
#include <unordered_set>
#include <deque>

class Solver
{
	std::filesystem::path filepath;

	void skip_spaces(std::ifstream &file)
	{
		while (file.peek() == ' ')
			file.get();
	}

	std::pair<std::unordered_set<std::string>, std::vector<std::string>> parse_one_line(std::ifstream &file)
	{
		assert(!file.eof());

		// Discard everything in the line up to ':'
		for (char c = file.get(); c != EOF && c != ':'; c = file.get())
		{
		}

		// Winning numbers
		std::unordered_set<std::string> winning_nums;
		std::string num;
		for (char c = file.get(); c != EOF && c != '|'; c = file.get())
		{
			if (c == ' ')
			{
				if (!num.empty())
				{
					winning_nums.insert(num);
				}
				num.clear();
				continue;
			}

			num += c;
		}
		if (!num.empty())
			winning_nums.insert(num);

		// Gotten numbers
		std::vector<std::string> gotten_numbers;
		num.clear();
		for (char c = file.get(); c != EOF && c != '\n'; c = file.get())
		{
			if (c == ' ')
			{
				if (!num.empty())
				{
					gotten_numbers.push_back(num);
				}
				num.clear();
				continue;
			}

			num += c;
		}
		if (!num.empty())
			gotten_numbers.push_back(num);

		return std::make_pair(winning_nums, gotten_numbers);
	}

	int part_1()
	{
		std::ifstream file;
		file.open(this->filepath);

		int result = 0;
		while (!file.eof())
		{
			int line_score = 0;
			auto [winning_numbers, gotten_numbers] = this->parse_one_line(file);
			for (const std::string &num : gotten_numbers)
			{
				if (!winning_numbers.contains(num))
					continue;

				line_score = line_score == 0 ? 1 : line_score * 2;
			}
			result += line_score;
		}

		file.close();
		return result;
	}

	int part_2()
	{
		std::ifstream file;
		file.open(this->filepath);

		int result = 0;
		std::deque<int> extra_cards;
		while (!file.eof())
		{
			auto [winning_numbers, gotten_numbers] = this->parse_one_line(file);
			int current_number_cards = extra_cards.empty() ? 1 : extra_cards.front() + 1;
			if (!extra_cards.empty())
				extra_cards.pop_front();

			int number_matching = 0;
			for (const std::string &num : gotten_numbers)
			{
				if (!winning_numbers.contains(num))
					continue;

				++number_matching;
			}

			for (int i = 0; i < number_matching; ++i)
			{
				if (extra_cards.size() == i)
				{
					extra_cards.push_back(current_number_cards);
				}
				else
				{
					extra_cards[i] += current_number_cards;
				}
			}
			result += number_matching * current_number_cards + 1;
		}

		file.close();
		return result;
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 4" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}

File: Day 4/solution.py
"""
Day 4 solution.
"""
import sys
from collections import deque


class Solver:
    """Day 4 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def parse_line(self, line: str) -> tuple[set[str], list[str]]:
        """
        Parse line for the winning numbers and the received numbers
        """
        game_details = line.split(":")[1]
        winning_nums, received_nums = game_details.split("|")
        return set(winning_nums.split()), received_nums.split()

    def part_1(self):
        """Part 1 solver."""

        def score_line(line: str) -> int:
            result = 0
            winning_numbers, gotten_numbers = self.parse_line(line)
            for num in gotten_numbers:
                if num not in winning_numbers:
                    continue

                if result == 0:
                    result = 1
                else:
                    result *= 2
            return result

        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            return sum(score_line(line) for line in file)

    def part_2(self):
        """Part 2 solver."""

        def calculate_extra_cards(line: str, extra_cards: deque) -> int:
            winning_numbers, gotten_numbers = self.parse_line(line)
            current_number_cards = (
                0 if not extra_cards else extra_cards.popleft()
            ) + 1

            number_matching = sum(
                1 for num in gotten_numbers if num in winning_numbers
            )

            for i in range(number_matching):
                if len(extra_cards) == i:
                    extra_cards.append(current_number_cards)
                else:
                    extra_cards[i] += current_number_cards

            return number_matching * current_number_cards + 1

        extra_cards = deque([])
        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            return sum(
                calculate_extra_cards(line, extra_cards) for line in file
            )

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 4")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case other:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()
